### PR TITLE
Add support for deepspeed optimizer and custom scheduler

### PR DESCRIPTION
**Pull request by pacman100**
# What does this PR do?
1. Add support for deepspeed optimizer and custom scheduler
2. Fixing the lr scheduler not being saved by passing them to the DeepSpeed engineer for all schedulers that are instances of LRScheduler.
3. Added related tests

### CODE REVIEW

File: src/accelerate/utils/deepspeed.py [hunk 1]
```diff
@@ -258,12 +258,15 @@ class DummyScheduler:
             Total number of steps.
         warmup_num_steps (int):
             Number of steps for warmup.
+        lr_scheduler_callable (callable):
```
Comment (BenjaminBossan) [inline src/accelerate/utils/deepspeed.py:261]: Should add "optional".

File: tests/deepspeed/test_deepspeed.py [hunk 2]
```diff
@@ -483,6 +483,31 @@ def test_prepare_deepspeed(self, optim_type, scheduler_type):
...
+                    model, optimizer, train_dataloader, eval_dataloader, lr_scheduler = accelerator.prepare(
+                        model, dummy_optimizer, train_dataloader, eval_dataloader, dummy_lr_scheduler
+                    )
+                self.assertTrue(
+                    "Either specify a scheduler in the config file or "
+                    "pass in the `lr_scheduler_callable` parameter when using `accelerate.utils.DummyScheduler`."
+                    in str(cm.exception)
+                )
+
+                # passing `lr_scheduler_callable` to DUmmyScheduler should enable DS Optim + Custom Scheduler
```
Comment (BenjaminBossan) [inline tests/deepspeed/test_deepspeed.py:497]:
```suggestion
                # passing `lr_scheduler_callable` to DummyScheduler should enable DS Optim + Custom Scheduler
```

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the PR. I'm probably missing the bigger picture, but where is `lr_scheduler_callable` actually called?

**Review by muellerzr — APPROVED** [review]
Great!
